### PR TITLE
MODLD-346 fix title source

### PR DESCRIPTION
**Pull request by aleksei-pronichev**
-add functionality for comparing dataField in post processors -mapper refactoring

### CODE REVIEW

File: src/main/java/org/folio/marc4ld/service/ld2marc/processing/combine/impl/AbstractDataFieldCombiner.java [hunk 1]
```diff
@@ -34,21 +41,36 @@ public Collection<DataField> build() {
...
+  private void putNonRepeatableFields(DataFieldContainer container) {
     getNonRepeatableFields()
       .stream()
-      .filter(this::notContainsInCombined)
-      .forEach(fieldTag -> putFieldIfPresent(fieldTag, dataField));
+      .filter(c -> isSuitForCombined(c, container))
+      .forEach(fieldTag -> replaceField(fieldTag, container));
+  }
+
+  private boolean isSuitForCombined(char fieldTag, DataFieldContainer dataFieldContainer) {
```
Comment (askhat-abishev) [inline src/main/java/org/folio/marc4ld/service/ld2marc/processing/combine/impl/AbstractDataFieldCombiner.java:71]: isSuitableForCombined

File: src/main/java/org/folio/marc4ld/service/ld2marc/processing/combine/impl/AbstractDataFieldCombiner.java [hunk 2]
```diff
@@ -58,6 +80,28 @@ private boolean notContainsInCombined(char fieldTag) {
       .isEmpty();
   }
 
+  private boolean hasPrioryFromExist(char fieldTag, ResourceEdge resourceEdge) {
```
Comment (askhat-abishev) [inline src/main/java/org/folio/marc4ld/service/ld2marc/processing/combine/impl/AbstractDataFieldCombiner.java:83]: hasGreaterPriorityFromExisting

**Review by askhat-abishev — APPROVED** [review]

File: src/main/java/org/folio/marc4ld/service/ld2marc/processing/combine/impl/AbstractDataFieldCombiner.java [hunk 3]
```diff
@@ -34,21 +41,36 @@ public Collection<DataField> build() {
...
     putRepeatableFields(dataField);
     setIndicators(dataField);
   }
 
-  private void putNonRepeatableFields(DataField dataField) {
+  protected abstract Collection<Character> getNonRepeatableFields();
+
+  protected abstract Collection<Character> getRepeatableFields();
+
+  protected Comparator<ResourceEdge> getComparator(char fieldTag) {
```
Comment (askhat-abishev) [inline src/main/java/org/folio/marc4ld/service/ld2marc/processing/combine/impl/AbstractDataFieldCombiner.java:60]: fieldTag is not being used

File: src/main/java/org/folio/marc4ld/service/ld2marc/processing/combine/impl/AbstractDataFieldCombiner.java [hunk 4]
```diff
@@ -58,6 +80,28 @@ private boolean notContainsInCombined(char fieldTag) {
       .isEmpty();
   }
 
+  private boolean hasPrioryFromExist(char fieldTag, ResourceEdge resourceEdge) {
+    var currentEdge = containedFields.getOrDefault(fieldTag, startedEdge);
+    return getComparator(fieldTag)
+      .compare(resourceEdge, currentEdge) > 0;
+  }
+
+  private void replaceField(char fieldTag, DataFieldContainer dataFieldContainer) {
```
Comment (askhat-abishev) [inline src/main/java/org/folio/marc4ld/service/ld2marc/processing/combine/impl/AbstractDataFieldCombiner.java:89]: replaceSubfield (looks like same for the method below)

**Review by PBobylev — CHANGES_REQUESTED** [review]
Target this PR to a 'release/alpha' branch please

File: src/main/java/org/folio/marc4ld/service/ld2marc/field/impl/Bibframe2MarcFieldRuleImpl.java [hunk 5]
```diff
@@ -57,9 +57,13 @@ public Bibframe2MarcFieldRuleImpl(
...
-  public boolean isSuitable(Resource resource, PredicateDictionary predicate) {
-    return Objects.equals(fieldRule.getTypes(), resource.getTypeNames())
-      && (isNull(predicate) || predicate.name().equals(fieldRule.getPredicate()));
+  public boolean isSuitable(ResourceEdge edge) {
+    var source = edge.getSource();
+    var target = edge.getTarget();
+    var predicate = edge.getPredicate();
+    return Objects.equals(fieldRule.getTypes(), target.getTypeNames())
+      && (isNull(predicate) || predicate.name().equals(fieldRule.getPredicate()))
+      && (isTypeLikeRuleParent(source) || isTypeLikeRuleParent(target));
```
Comment (pkjacob) [inline src/main/java/org/folio/marc4ld/service/ld2marc/field/impl/Bibframe2MarcFieldRuleImpl.java:66]: `|| isTypeLikeRuleParent(target)` is not needed here right? 
We need to compare the type of the source only with field rule's parent.

File: src/main/java/org/folio/marc4ld/service/ld2marc/resource/Resource2MarcRecordMapperImpl.java [hunk 6]
```diff
@@ -50,22 +52,36 @@ public Record toMarcRecord(Resource resource) {
...
-      var dataFields = rules.stream()
-        .flatMap(tagToRule -> mapToDataFields(resource, predicate, cfb, tagToRule));
-
-      return Stream.concat(dataFields,
-          resource.getOutgoingEdges()
-            .stream()
-            .flatMap(re -> getFields(re.getTarget(), re.getPredicate(), cfb).stream()))
-        .toList();
-    }
+  private List<DataFieldContainer> getFields(ResourceEdge edge, ControlFieldsBuilder cfb) {
```
[the remark this answers is not in the document]
Comment (aleksei-pronichev) [reply src/main/java/org/folio/marc4ld/service/ld2marc/resource/Resource2MarcRecordMapperImpl.java]: container was removed

File: src/test/java/org/folio/marc4ld/mapper/Bibframe2Marc245IT.java [hunk 7]
```diff
@@ -126,4 +140,42 @@ private Resource createResourceWithWorkWith245() {
       outgoingResources
     );
   }
+
+  private Resource createResourceWithWorkAndInstanceSimilarTitles() {
```
[the remark this answers is not in the document]
Comment (aleksei-pronichev) [reply src/test/java/org/folio/marc4ld/mapper/Bibframe2Marc245IT.java:144]: container was removed

File: src/main/java/org/folio/marc4ld/service/ld2marc/resource/Resource2MarcRecordMapperImpl.java [hunk 8]
```diff
@@ -50,22 +52,36 @@ public Record toMarcRecord(Resource resource) {
...
+    var dataField = ld2MarcMapper.map(edge.getTarget());
+    var container = new DataFieldContainer(edge, dataField);
+    return List.of(container);
+
+  }
+
+  public List<DataFieldContainer> getDataFieldContainer(ControlFieldsBuilder cfb, ResourceEdge edge) {
+    var resource = edge.getTarget();
+    var predicate = edge.getPredicate();
+    var dataFields = rules.stream()
```
[the remark this answers is not in the document]
Comment (aleksei-pronichev) [reply src/main/java/org/folio/marc4ld/service/ld2marc/resource/Resource2MarcRecordMapperImpl.java:64]: container was removed

File: src/main/java/org/folio/marc4ld/service/ld2marc/processing/combine/impl/DataFieldCombinerFactoryImpl.java [hunk 9]
```diff
@@ -24,8 +25,8 @@ private static final class DefaultDataFieldCombiner implements DataFieldCombiner
     private final List<DataField> fields = new ArrayList<>();
 
     @Override
-    public void add(DataField dataField) {
-      fields.add(dataField);
+    public void add(DataFieldContainer dataField) {
```
[the remark this answers is not in the document]
Comment (aleksei-pronichev) [reply src/main/java/org/folio/marc4ld/service/ld2marc/processing/combine/impl/DataFieldCombinerFactoryImpl.java]: container was removed

File: src/main/java/org/folio/marc4ld/service/ld2marc/processing/combine/impl/AbstractDataFieldCombiner.java [hunk 10]
```diff
@@ -8,21 +8,28 @@
...
 import org.marc4j.marc.Subfield;
 
 public abstract class AbstractDataFieldCombiner implements DataFieldCombiner {
 
-  private DataField combinedField;
+  private final Comparator<ResourceEdge> defaultComparator = new InstancePriorityComparator();
 
-  protected abstract Collection<Character> getNonRepeatableFields();
+  private DataField combinedField;
+  private ResourceEdge startedEdge;
```
[the remark this answers is not in the document]
Comment (aleksei-pronichev) [reply src/main/java/org/folio/marc4ld/service/ld2marc/processing/combine/impl/AbstractDataFieldCombiner.java]: approach was reverted

File: src/main/java/org/folio/marc4ld/service/ld2marc/processing/combine/impl/AbstractDataFieldCombiner.java (hunk 1 above)
Comment (aleksei-pronichev) [reply src/main/java/org/folio/marc4ld/service/ld2marc/processing/combine/impl/AbstractDataFieldCombiner.java]: rollback approach

File: src/main/java/org/folio/marc4ld/service/ld2marc/processing/combine/impl/AbstractDataFieldCombiner.java (hunk 3 above)
Comment (aleksei-pronichev) [reply src/main/java/org/folio/marc4ld/service/ld2marc/processing/combine/impl/AbstractDataFieldCombiner.java]: rollback approach

File: src/main/java/org/folio/marc4ld/service/ld2marc/processing/combine/impl/AbstractDataFieldCombiner.java (hunk 2 above)
Comment (aleksei-pronichev) [reply src/main/java/org/folio/marc4ld/service/ld2marc/processing/combine/impl/AbstractDataFieldCombiner.java]: rollback approach

File: src/main/java/org/folio/marc4ld/service/ld2marc/processing/combine/impl/AbstractDataFieldCombiner.java (hunk 4 above)
Comment (aleksei-pronichev) [reply src/main/java/org/folio/marc4ld/service/ld2marc/processing/combine/impl/AbstractDataFieldCombiner.java]: rollback approach

File: src/main/java/org/folio/marc4ld/service/ld2marc/field/impl/Bibframe2MarcFieldRuleImpl.java (hunk 5 above)
Comment (aleksei-pronichev) [reply src/main/java/org/folio/marc4ld/service/ld2marc/field/impl/Bibframe2MarcFieldRuleImpl.java:66]: Thanks! 
There was a misunderstanding of the setting

**Review by PBobylev — APPROVED** [review]

**Review by pkjacob — APPROVED** [review]

**Pull request merged**